README
SMS-Send-Ipipi version 0.01
===========================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  blah blah blah

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2008 by Andrew Moore

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension SMS::Send::US::Ipipi.

0.02 Mon June 1 2008
    - fixed test suite to deal with login information

0.01  Tue May 27 13:35:03 2008
    - original version; created by h2xs 1.23 with options
        -AX -n SMS::Send::US::Ipipi
    - copied Adam Kennedy's SMS::Send::AU::MyVodafone module.



lib/SMS/Send/Template.pm
package SMS::Send::Ipipi;

use 5.008008;
use strict;
use warnings;

require Exporter;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use SMS::Send::Ipipi ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);

our $VERSION = '0.01';


# Preloaded methods go here.

1;
__END__
# Below is stub documentation for your module. You'd better edit it!

=head1 NAME

SMS::Send::Ipipi - Perl extension for blah blah blah

=head1 SYNOPSIS

  use SMS::Send::Ipipi;
  blah blah blah

=head1 DESCRIPTION

Stub documentation for SMS::Send::Ipipi, created by h2xs. It looks like the
author of the extension was negligent enough to leave the stub
unedited.

Blah blah blah.

=head2 EXPORT

None by default.



=head1 SEE ALSO

Mention other useful documentation such as the documentation of
related modules or operating system documentation (such as man pages
in UNIX), or any relevant external documentation such as RFCs or
standards.

If you have a mailing list set up for your module, mention it here.

If you have a web site set up for your module, mention it here.

=head1 AUTHOR

Andrew Moore, E<lt>acm@E<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2008 by Andrew Moore

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.


=cut


lib/SMS/Send/MyVodafone.pm
package SMS::Send::AU::MyVodafone;

=pod

=head1 NAME

SMS::Send::AU::MyVodafone - An SMS::Send driver for the myvodafone.com.au website

=head1 SYNOPSIS

  # Get the sender and login
  my $sender = SMS::Send->new('AU::MyVodafone',
  	_login    => '04 444 444', # Whitespace is ignored
  	_password => 'abcdefg',
  	);
  
  # Send a message to ourself
  my $sent = $sender->send_sms(
  	text => 'Messages have a limit of 160 chars',
  	to   => '+61 4 444 444',
  	);
  
  # Did it send?
  if ( $sent ) {
  	print "Sent test message\n";
  } else {
  	print "Test message failed\n";
  }

=head1 DESCRIPTION

L<SMS::Send::AU::MyVodaphone> is an regional L<SMS::Send> driver for
Australia that delivers messages via the L<http://myvodafone.com.au>
website Web2TXT feature.

Using your phone number as a login, and your existing password, this
driver allows any Australian with a Vodafone to send SMS messages (with
the message cost added to your account).

=head2 Preparing to Use This Driver

As well as setting up your myvodfone.com.au account and password, the
Web2TXT feature requires acceptance of an additional disclaimer and
conditions of use form.

You C<must> manually accept this disclaimer and conditions before you
will be able to use this driver.

While we certainly could make the driver do it for you, acceptance
of the terms of use implies you understand the cost structure and
rules surrounding the use of the Web2TXT feature.

=head2 Disclaimer

Other than dieing on encountering the terms of use form, no other
protection is provided, and the authors of this driver take no
responsibility for any costs accrued on your phone bill by using
this module.

Using this driver will cost you money. B<YOU HAVE BEEN WARNED>

=head1 METHODS

=cut

use 5.005;
use strict;
use base 'SMS::Send::Driver';
use WWW::Mechanize ();

use vars qw{$VERSION};
BEGIN {
	$VERSION = '0.03';
}

# Starting URI
my $START = 'https://www.myvodafone.com.au/knox/login_handler.jsp';
my $FORM  = 'https://www.myvodafone.com.au/yrweb2txt/enter.do';

# Detection regexs
my $RE_BADLOGIN = qr/Sorry you must enter a valid username and password/i;





#####################################################################
# Constructor

=pod

=head2 new

  # Create a new sender using this driver
  my $sender = SMS::Send->new(
  	_login    => '04 444 444',
  	_password => 'abcdefg',
  	);

The C<new> constructor takes two parameters, which should be passed
through from the L<SMS::Send> constructor.

The params are driver-specific for now, until L<SMS::Send> adds a standard
set of params for specifying the login and password.

=over

=item _login

The C<_login> param should be your phone number. That is, the phone to send
from and to be billed to for the messages.

The login should be an Australian-format number. That is, starting with
zero-four "04".

=item _password

The C<_password> param should be your myvodafone.com.au password.

=back

During construction, the driver will actively log in to the
myvodafone.com.au website using the credentials provided, to verify
they are correct.

It will not check for your acceptance of the terms and conditions at
this point. That is done at C<send_sms>-time.

Returns a new C<SMS::Send::AU::MyVodafone> object, or dies on error.

=cut

sub new {
	my $class  = shift;
	my %params = @_;

	# Get the login
	my $login    = $class->_LOGIN   ( delete $params{_login}    );
	my $password = $class->_PASSWORD( delete $params{_password} );

	# Create our mechanise object
	my $mech = WWW::Mechanize->new;

	# Create the object, saving any private params for later
	my $self = bless {
		mech     => $mech,
		login    => $login,
		password => $password,
		private  => \%params,

		# State variables
		logged_in => '',
		}, $class;

	$self;
}

sub _get_login {
	my $self = shift;

	# Get to the login form
	$self->{mech}->get( $START );
	unless ( $self->{mech}->success ) {
		Carp::croak("HTTP Error: Failed to connect to MyVodafone website");
	}

	return 1;
}

sub _send_login {
	my $self = shift;

	# Shortcut if logged in
	return 1 if $self->{logged_in};

	# Get to the login page
	$self->_get_login;

	# Submit the login form
	$self->{mech}->submit_form(
		form_name => 'loginForm',
		fields    => {
			txtUserID   => $self->{login},
			txtPassword => $self->{password},
			btnLogin    => 'submit',
			},
		);

	# Did we login?
	if ( $self->{mech}->content =~ $RE_BADLOGIN ) {
		Carp::croak("Invalid login and password");
	}

	$self->{logged_in} = 1;
	return 1;
}

sub send_sms {
	my $self   = shift;
	my %params = @_;

	# Get the message and destination
	my $message   = $self->_MESSAGE( delete $params{text} );
	my $recipient = $self->_TO     ( delete $params{to}   );

	# Make sure we are logged in
	$self->_send_login;

	# Get to the Web2TXT form
	$self->{mech}->get( $FORM );
	unless ( $self->{mech}->content =~ /Compose a message/ ) {
		Carp::croak("Could not locate the SMS send form");
	}

	# Fill out the message form
	my $form = $self->{mech}->form_name('sendMessageForm')
		or Carp::croak("Failed to find sendMessageForm on message page");
	$form->push_input('text', {
		name  => 'recipients',
		value => "adhoc$recipient",
		} );
	$form->value( messageBody => $message               );

	# Hack some values otherwise changed by JavaScript.
	# Disable warnings when changing hidden inputs.
	SCOPE: {
		local $^W = 0;
		$form->value( action      => 'send'                 );
		$form->value( counter     => 160 - length($message) );
		$form->value( msg_counter => 1                      );
		$form->value( totalMsgs   => 1                      );
	}

	# Send the form
	$self->{mech}->submit();
	unless ( $self->{mech}->success ) {
		Carp::croak("HTTP request returned failure when sending SMS request");
	}

	# Fire-and-forget, we don't know for sure.
	return 1;
}





#####################################################################
# Support Functions

sub _LOGIN {
	my $class  = ref $_[0] ? ref shift : shift;
	my $number = shift;
	unless ( defined $number and ! ref $number and length $number ) {
		Carp::croak("Did not provide a login number");
	}
	$number =~ s/\s//g;
	unless ( $number =~ /^04\d{8}$/ ) {
		Carp::croak("Login must be a 10-digit number starting with 04");
	}
	return $number;
}

sub _PASSWORD {
	my $class    = ref $_[0] ? ref shift : shift;
	my $password = shift;
	unless ( defined $password and ! ref $password and length $password ) {
		Carp::croak("Did not provide a password");
	}
	unless ( length($password) >= 5 ) {
		Carp::croak("Password must be at least 5 characters");
	}
	return $password;
}

sub _MESSAGE {
	my $class   = ref $_[0] ? ref shift : shift;
	my $message = shift;
	unless ( length($message) <= 160 ) {
		Carp::croak("Message length limit is 160 characters");
	}
	return $message;
}

sub _TO {
	my $class = ref $_[0] ? ref shift : shift;
	my $to    = shift;

	# International numbers need their + removed
	if ( $to =~ s/^\+// ) {
		return $to;
	}

	# Domestic numbers start with 04
	unless ( $to =~ /^04\d{8}$/ ) {
		Carp::croak("Regional number is not an Australian mobile phone number");
	}

	return $to;
}

1;

=pod

=head1 SUPPORT

Bugs should be reported via the CPAN bug tracker at

L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=SMS-Send-AU-MyVodafone>

For other issues, contact the author.

=head1 AUTHOR

Adam Kennedy E<lt>adamk@cpan.orgE<gt>, L<http://ali.as/>

=head1 COPYRIGHT

Copyright 2005 Adam Kennedy.

This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.

Additionally, you are again reminded that this software comes with
no warranty of any kind, including but not limited to the implied
warranty of merchantability.

ANY use my result in charges on your phone bill, and you should use this
software with care. The author takes no responsibility for any such
charges accrued.

=cut


lib/SMS/Send/US/Ipipi.pm
package SMS::Send::US::Ipipi;
use warnings;
use strict;

=head1 NAME

SMS::Send::US::Ipipi - An SMS::Send driver for the ipipi.com website

=head1 VERSION

Version 0.02

=cut

our $VERSION = '0.02';

=head1 SYNOPSIS

  # Get the sender and login
  my $sender = SMS::Send->new('US::Ipipi',
  	_login    => 'username',
  	_password => 'password',
  	);
  
  my $sent = $sender->send_sms(
  	text => 'Messages have a limit of 160 chars',
  	to   => '212-555-1212',
  	);
  
  # Did it send?
  if ( $sent ) {
  	print "Sent test message\n";
  } else {
  	print "Test message failed\n";
  }

=head1 DESCRIPTION

L<SMS::Send::US::Ipipi> is a L<SMS::Send> driver that delivers messages
via the L<http://ipipi.com> website.

=head2 Preparing to Use This Driver

You need to sign up for an account at http://ipipi.com to be able to
use this driver.

=head2 Disclaimer

Using this driver may cost you money. B<YOU HAVE BEEN WARNED>

=head1 METHODS

=cut

use base 'SMS::Send::Driver';
use LWP::UserAgent ();


=head2 new

  # Create a new sender using this driver
  my $sender = SMS::Send->new(
  	_login    => 'username',
  	_password => 'password',
  	);

The C<new> constructor takes two parameters, which should be passed
through from the L<SMS::Send> constructor.

The params are driver-specific for now, until L<SMS::Send> adds a standard
set of params for specifying the login and password.

=over

=item _login

The C<_login> param should be your ipipi.com login.

=item _password

The C<_password> param should be your ipipi.com password.

=back

Returns a new C<SMS::Send::US::Ipipi> object, or dies on error.

=cut

sub new {
	my $class  = shift;
	my %params = @_;

	# Get the login
	my $login    = $class->_LOGIN   ( delete $params{_login}    );
	my $password = $class->_PASSWORD( delete $params{_password} );

        # Create our LWP::UserAgent object
        my $ua = LWP::UserAgent->new;

	# Create the object, saving any private params for later
	my $self = bless {
                          ua       => $ua,
                          login    => $login,
                          password => $password,
                          private  => \%params,
                          
                          # State variables
                          logged_in => '',
                      }, $class;

	$self;
}

=head2 send_sms

  This method is actually called by SMS::Send when you call send_sms on it.

  my $sent = $sender->send_sms(
  	text => 'Messages have a limit of 160 chars',
  	to   => '212-555-1212',
  	);

=cut

sub send_sms {
    my $self   = shift;
    my %params = @_;
    
    # Get the message and destination
    my $message   = $self->_MESSAGE( delete $params{text} );
    my $recipient = $self->_TO     ( delete $params{to}   );

    my $response = $self->{'ua'}->post( 'http://service.ipipi.com/wsrv/api.asmx/xmlSendSMS',
                                       { Username => $self->{'login'},
                                         Password => $self->{'password'},
                                         SendTo   => $recipient,
                                         Text     => $message,
                                         Encoding => 7, }, );
    
    
    if ( not $response->is_success() ) {
        Carp::croak( 'HTTP request returned failure when sending SMS request: ' . $response->status_line() );
    }

    # warn $response->content();
    
    # Fire-and-forget, we don't know for sure.
    return 1;
}


#####################################################################
# Support Functions

sub _LOGIN {
	my $class  = ref $_[0] ? ref shift : shift;
	my $login = shift;
	unless ( defined $login and ! ref $login and length $login ) {
		Carp::croak('Did not provide a login');
	}
	return $login;
}

sub _PASSWORD {
	my $class    = ref $_[0] ? ref shift : shift;
	my $password = shift;
	unless ( defined $password and ! ref $password and length $password ) {
		Carp::croak("Did not provide a password");
	}
	return $password;
}

sub _MESSAGE {
	my $class   = ref $_[0] ? ref shift : shift;
	my $message = shift;
	unless ( length($message) <= 160 ) {
		Carp::croak("Message length limit is 160 characters");
	}
	return $message;
}

sub _TO {
	my $class = ref $_[0] ? ref shift : shift;
	my $to    = shift;

        # strip out non-numerals
        $to =~ s/[^\d]//g;

	# International numbers need their + removed
	if ( $to =~ s/^\+// ) {
		return $to;
	}
        
        if ( $to !~ /^1/ ) {
            $to = '1' . $to;
        }
        
 	# US numbers should be 11 digits, starting with "1"
 	unless ( $to =~ /^1\d{10}$/ ) {
 		Carp::croak("Regional number is not a valid US mobile phone number");
 	}

	return $to;
}

1;

=head1 AUTHOR

Andrew Moore, C<< <andrew.moore at liblime.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-sms-send-ipipi at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=SMS-Send-US-Ipipi>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc SMS::Send::US::Ipipi

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=SMS-Send-US-Ipipi>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/SMS-Send-US-Ipipi>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/SMS-Send-US-Ipipi>

=item * Search CPAN

L<http://search.cpan.org/dist/SMS-Send-US-Ipipi>

=back


=head1 ACKNOWLEDGEMENTS

Thanks to Adam Kennedy E<lt>adamk@cpan.orgE<gt>, L<http://ali.as/> for
writing SMS::Send and for SMS::Send::AU::MyVodafone which I copied for
this module.

=head1 COPYRIGHT & LICENSE

Copyright (C) 2008 LibLime L<http://liblime.com>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.

Additionally, you are again reminded that this software comes with
no warranty of any kind, including but not limited to the implied
warranty of merchantability.

ANY use my result in charges on your ipipi.com bill, and you should
use this software with care. The author takes no responsibility for
any such charges accrued.

=cut

1; # End of SMS::Send::US::Ipipi

